### DISCUSSION TITLE
Redundant cast needed from `list` to `list[object]`?

### DISCUSSION THREAD

**Discussion by Holt59** [microsoft/pyright#5979] [Q&A]
I am trying to make a `TypeGuard` that checks if an object `e` is a list containing only instances of a class `Function`:

```
def is_list_of_functions(e: object) -> TypeGuard[list[Function]]:
    if not isinstance(e, list):
        return False
    return all(isinstance(x, Function) for x in e)
```

This works, but `pyright` warns about the type of `x` being "unknown", which is kind of right but not really relevant here... I can cast `e` to `list[object]` which removes the issue, but this feels very redundant, and I cannot check `isinstance(e, list[object])`. Is there a clean way to achieve this without requiring a `cast` on `e`?

**Comment by erictraut** [comment]
I presume that you've chosen to explicitly enable the `reportUnknownVariableType` diagnostic check — or you've enabled "strict" type checking mode, which implicitly enables `reportUnknownVariableType`. This check reports any case where a variable is assigned a value whose type is `Unknown` or contains an `Unknown`. 

`Unknown` is the name that pyright uses for an "implicit `Any`" — i.e. an `Any` type that was not specified explicitly in the code. In your example, the type of `e` is narrowed to `list[Unknown]` by the `isinstance` type guard, and the `for` statement assigns an `Unknown` to the variable `x`.

If you don't care about the `reportUnknownVariableType` check, you can refrain from enabling it. By default, it's off.

Another workaround is to annotate the parameter `e` with the type `Any | Iterable[Any]`. This happens to avoid producing an `Unknown` is pyright's current `isinstance` type narrowing logic.

A `cast` may be needed in some cases to work around issues like this.

**Reply by Holt59** [reply]
Thanks, I have `strict` mode enabled for so long in VS Code that I did not think of mentioning it... The `Any | Iterable[Any]` looks like a good solution to me (and works).
